src/peer_map.rs

```diff
@@ -40,172 +35,100 @@ impl PeerMap {
-        conn_info: ConnectionInfo,
-    ) -> Result<PublicId, ConnectionError> {
-        match self.network_to_routing.entry(conn_info.peer_addr()) {
-            Entry::Vacant(entry) => {
-                let _ = entry.insert(State::Network(PeerType::from(conn_info)));
-                Err(ConnectionError::Incomplete)
-            }
-            Entry::Occupied(mut entry) => match *entry.get() {
-                State::Network(_) => Err(ConnectionError::Incomplete),
-                State::Routing(pub_id) => {
-                    *entry.get_mut() = State::Complete(pub_id);
-                    Ok(pub_id)
-                }
-                State::Complete(_) => Err(ConnectionError::AlreadyConnected),
-            },
+    // TODO: remove this `allow` when https://github.com/rust-lang/rust-clippy/issues/4219
+    // is fixed.
+    #[allow(clippy::map_entry)]
+    pub fn handle_connected_to(&mut self, conn_info: ConnectionInfo) {
+        let socket_addr = conn_info.peer_addr();
+        if !self.reverse.contains_key(&socket_addr) {
+            let _ = self
+                .pending
+                .insert(socket_addr, PendingConnection::from(conn_info));
         }
     }
 
     // Handles `ConnectionFailure` event from the network layer. Returns the `PublicId` of the peer,
     // if the connection to them was previously established. Otherwise returns `None`.
     pub fn handle_connection_failure(&mut self, socket_addr: SocketAddr) -> Option<PublicId> {
-        match self.network_to_routing.remove(&socket_addr) {
-            Some(State::Complete(pub_id)) => {
-                let _ = self.routing_to_network.remove(&pub_id);
-                Some(pub_id)
-            }
-            Some(_) | None => None,
+        let _ = self.pending.remove(&socket_addr);
+
+        if let Some(pub_id) = self.reverse.remove(&socket_addr) {
+            let _ = self.forward.remove(&pub_id);
+            Some(pub_id)
+        } else {
+            None
         }
     }
 
     // Handles received `DirectMessage`.
-    pub fn handle_direct_message(
-        &mut self,
-        pub_id: PublicId,
-        socket_addr: SocketAddr,
-    ) -> Result<ConnectionInfo, ConnectionError> {
-        if let Some(routing_peer) = self.network_to_routing.get_mut(&socket_addr) {
-            match mem::replace(routing_peer, State::Complete(pub_id)) {
-                State::Network(peer) => {
-                    let conn_info = peer.into_connection_info(socket_addr);
-                    let _ = self.routing_to_network.insert(pub_id, conn_info.clone());
-                    Ok(conn_info)
-                }
-                State::Routing(_) => Self::invalid_state(),
-                State::Complete(_) => Err(ConnectionError::AlreadyConnected),
-            }
-        } else {
-            Self::invalid_state()
+    pub fn handle_direct_message(&mut self, pub_id: PublicId, socket_addr: SocketAddr) {
+        if let Some(peer_type) = self.pending.remove(&socket_addr) {
```

Comment:
This entire commits looks very good to me: lots of simplifications etc., clear abstractions. I'm just unclear on one thing in the new workflow: why does a Peer need to be in the `pending` map (i.e be pending connection) for us to handle a direct message from them? Shouldn't it also work if we find them in our `reverse` map, i.e: if we're already connected to them?
I may be confusing myself, but hopefully this explains what's unclear to me and you can help me understand.

Reply:
When someone whom we haven't established a connection yet sends us `DirectMessage`, the flow goes like this:
1. First quic-p2p emits `Event::ConnectedTo` which contains their connection info. We put it into `pending`.
2. Then quic-p2p emits `Event::NewMessage` which contains the socket address of the sender and the message itself (which if it is `DirectMessage` contains the public id of the sender). So we lookup the previously stored connection info in `pending` by the socket address and then store the connection in both `forward` and `reverse`.

It is necessary to do it like this, because `ConnectedTo` contains only the connection info and `NewMessage` contains only socket address (not the full connection info).

So when we receive a `DirectMessage` from someone for the first time, the above flow happens. When we receive any subsequent messages, there is no entry in `pending` for them anymore (because we removed it before), so there is nothing more to do. The full connection information is already stored in `forward` and `reverse`.

Now I'm thinking that this method (and the others too) might be poorly named as `PeerMap` doesn't really know or care about `DirectMessage`s. It only cares about the association of connection infos to public ids. I'll try to come up with better names.

Reply:
Would be good to put something along that information in the PR description.

Reply:
I can put it there, problem is this PR doesn't introduce this flow. It was already introduced in the previous PR (where it was actually explained in the PR description, although in less details). So putting it here might be a bit misleading? I think I'll rather expand the comment for `PeerMap`.

Reply:
It probably would not be misleading, but likely difficult to find. A code comment is a good idea (thought if you cannot put all the information there, PR description is still probably good.)

Reply:
Misleading in that it would suggest the change was introduced in this PR when in reality it wasn't. Probably not a major problem. Still, I'm putting it as a comment so it's easier to find.

Reply:
@madadam, I agree with the naming: really it's the name of the function that confused me here now you've explained it :)